extension/json/json_functions/json_transform.cpp

```diff
@@ -0,0 +1,375 @@
+#include "duckdb/common/types.hpp"
+#include "duckdb/function/scalar/nested_functions.hpp"
+#include "json_common.hpp"
+#include "json_functions.hpp"
+
+namespace duckdb {
+
+//! Forward declaration for recursion
+static LogicalType StructureToType(yyjson_val *val);
+
+static LogicalType StructureToTypeArray(yyjson_val *arr) {
+	if (yyjson_arr_size(arr) != 1) {
+		throw InvalidInputException("Too many values in array of JSON structure");
+	}
+	return LogicalType::LIST(StructureToType(yyjson_arr_get_first(arr)));
+}
+
+static LogicalType StructureToTypeObject(yyjson_val *obj) {
+	unordered_set<string> names;
+	child_list_t<LogicalType> child_types;
+	yyjson_val *key, *val;
+	yyjson_obj_iter iter;
+	yyjson_obj_iter_init(obj, &iter);
+	while ((key = yyjson_obj_iter_next(&iter))) {
+		val = yyjson_obj_iter_get_val(key);
+		auto key_str = yyjson_get_str(key);
+		if (names.find(key_str) != names.end()) {
+			auto obj_string = JSONCommon::WriteVal(obj);
+			throw InvalidInputException("Duplicate keys in object in JSON structure: %s", obj_string.GetString());
+		}
+		names.insert(key_str);
+		child_types.emplace_back(key_str, StructureToType(val));
+	}
+	return LogicalType::STRUCT(child_types);
+}
+
+static LogicalType StructureToTypeVal(yyjson_val *val) {
+	auto type_string = StringUtil::Lower(yyjson_get_str(val));
+	LogicalType result;
+	if ((StringUtil::StartsWith(type_string, "decimal(") || StringUtil::StartsWith(type_string, "dec(") ||
```

Comment:
Do we need these string operations for type comparisons here, or is there a different way of reading this?

Reply:
`TransformStringToLogicalType` returns  works for most logical types. The problem here is that `decimal` is parameterised. We do not have a standard way of parsing such strings into a logical type.

I could, for instance, change `TransformStringToLogicalType` to `TransformStringToLogicalTypeId`, and add a new function `TransformStringToLogicalType`, which can read parameterised types, and return an actual `LogicalType`, rather than `LogicalTypeId`. This is probably more future-proof.